<!DOCTYPE html>
<html lang="zh">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.4.0">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">



<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.15.4/css/all.min.css" integrity="sha256-mUZM63G8m73Mcidfrv5E+Y61y7a12O5mW4ezU3bxqW4=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/animate.css@3.1.1/animate.min.css" integrity="sha256-PR7ttpcvz8qrF57fur/yAx1qXMFJeJFiA6pSzWi0OIE=" crossorigin="anonymous">

<script class="next-config" data-name="main" type="application/json">{"hostname":"vampireqian.github.io","root":"/","images":"/images","scheme":"Muse","darkmode":false,"version":"8.8.2","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12},"copycode":false,"bookmark":{"enable":false,"color":"#222","save":"auto"},"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"stickytabs":false,"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"fadeInDown","post_body":"fadeInDown","coll_header":"fadeInLeft","sidebar":"fadeInUp"}},"prism":false,"i18n":{"placeholder":"بحث...","empty":"We didn't find any results for the search: ${query}","hits_time":"${hits} results found in ${time} ms","hits":"${hits} results found"}}</script><script src="/js/config.js"></script>
<meta name="description" content="异常处理异常Exception 错误 Error ：错误是可以避免的逻辑错误：算法写错了，加法写成了减法笔误：变量名写错了，语法错误函数或类使用错误，其实这也属于逻辑错误  异常 Exception ：异常不可能避免本意就是意外情况这有个前提，没有出现上面说的错误，也就是说程序写的没有问题，但是在某些情况下，会出现一些意外，导致程序无法正常的执行下去。例如open函数操作一个文件，文件不存在，或者">
<meta property="og:type" content="article">
<meta property="og:title" content="python-异常处理、模块化">
<meta property="og:url" content="https://vampireqian.github.io/2017/09/22/11python%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86%E3%80%81%E6%A8%A1%E5%9D%97%E5%8C%96/index.html">
<meta property="og:site_name" content="luoqian">
<meta property="og:description" content="异常处理异常Exception 错误 Error ：错误是可以避免的逻辑错误：算法写错了，加法写成了减法笔误：变量名写错了，语法错误函数或类使用错误，其实这也属于逻辑错误  异常 Exception ：异常不可能避免本意就是意外情况这有个前提，没有出现上面说的错误，也就是说程序写的没有问题，但是在某些情况下，会出现一些意外，导致程序无法正常的执行下去。例如open函数操作一个文件，文件不存在，或者">
<meta property="og:locale">
<meta property="article:published_time" content="2017-09-21T16:00:00.000Z">
<meta property="article:modified_time" content="2018-12-11T08:36:38.000Z">
<meta property="article:author" content="luoqian">
<meta property="article:tag" content="python">
<meta name="twitter:card" content="summary">


<link rel="canonical" href="https://vampireqian.github.io/2017/09/22/11python%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86%E3%80%81%E6%A8%A1%E5%9D%97%E5%8C%96/">



<script class="next-config" data-name="page" type="application/json">{"sidebar":"","isHome":false,"isPost":true,"lang":"zh","comments":true,"permalink":"https://vampireqian.github.io/2017/09/22/11python%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86%E3%80%81%E6%A8%A1%E5%9D%97%E5%8C%96/","path":"2017/09/22/11python异常处理、模块化/","title":"python-异常处理、模块化"}</script>

<script class="next-config" data-name="calendar" type="application/json">""</script>
<title>python-异常处理、模块化 | luoqian</title>
  




  <noscript>
    <link rel="stylesheet" href="/css/noscript.css">
  </noscript>
</head>

<body itemscope itemtype="http://schema.org/WebPage" class="use-motion">
  <div class="headband"></div>

  <main class="main">
    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="تشغيل شريط التصفح" role="button">
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <i class="logo-line"></i>
      <p class="site-title">luoqian</p>
      <i class="logo-line"></i>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>







</div>
        
  
  <div class="toggle sidebar-toggle" role="button">
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
  </div>

  <aside class="sidebar">

    <div class="sidebar-inner sidebar-nav-active sidebar-toc-active">
      <ul class="sidebar-nav">
        <li class="sidebar-nav-toc">
          المحتويات
        </li>
        <li class="sidebar-nav-overview">
          عام
        </li>
      </ul>

      <div class="sidebar-panel-container">
        <!--noindex-->
        <div class="post-toc-wrap sidebar-panel">
            <div class="post-toc animated"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86"><span class="nav-number">1.</span> <span class="nav-text">异常处理</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%BC%82%E5%B8%B8Exception"><span class="nav-number">1.1.</span> <span class="nav-text">异常Exception</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E4%BA%A7%E7%94%9F%E5%BC%82%E5%B8%B8"><span class="nav-number">1.2.</span> <span class="nav-text">产生异常</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%BC%82%E5%B8%B8%E7%B1%BB%E5%8F%8A%E7%BB%A7%E6%89%BF%E5%B1%82%E6%AC%A1"><span class="nav-number">1.3.</span> <span class="nav-text">异常类及继承层次</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#BaseException%E5%8F%8A%E5%AD%90%E7%B1%BB"><span class="nav-number">1.4.</span> <span class="nav-text">BaseException及子类</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#Exception%E5%8F%8A%E5%AD%90%E7%B1%BB"><span class="nav-number">1.5.</span> <span class="nav-text">Exception及子类</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E6%9C%AA%E5%AE%9E%E7%8E%B0%E5%92%8C%E6%9C%AA%E5%AE%9E%E7%8E%B0%E5%BC%82%E5%B8%B8"><span class="nav-number">1.6.</span> <span class="nav-text">未实现和未实现异常</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%BC%82%E5%B8%B8%E7%9A%84%E6%8D%95%E8%8E%B7"><span class="nav-number">1.7.</span> <span class="nav-text">异常的捕获</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#finally%E5%AD%90%E5%8F%A5"><span class="nav-number">1.8.</span> <span class="nav-text">finally子句</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%BC%82%E5%B8%B8%E7%9A%84%E4%BC%A0%E9%80%92"><span class="nav-number">1.9.</span> <span class="nav-text">异常的传递</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#try%E5%B5%8C%E5%A5%97"><span class="nav-number">1.10.</span> <span class="nav-text">try嵌套</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%BC%82%E5%B8%B8%E7%9A%84%E6%8D%95%E8%8E%B7%E7%9A%84%E6%97%B6%E6%9C%BA"><span class="nav-number">1.11.</span> <span class="nav-text">异常的捕获的时机</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#else%E5%AD%90%E5%8F%A5"><span class="nav-number">1.12.</span> <span class="nav-text">else子句</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#%E6%80%BB%E7%BB%93"><span class="nav-number">1.13.</span> <span class="nav-text">总结</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E6%A8%A1%E5%9D%97%E5%8C%96"><span class="nav-number">2.</span> <span class="nav-text">模块化</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#%E5%AF%BC%E5%85%A5%E8%AF%AD%E5%8F%A5"><span class="nav-number">2.1.</span> <span class="nav-text">导入语句</span></a></li></ol></li></ol></div>
        </div>
        <!--/noindex-->

        <div class="site-overview-wrap sidebar-panel">
          <div class="site-author site-overview-item animated" itemprop="author" itemscope itemtype="http://schema.org/Person">
  <p class="site-author-name" itemprop="name">luoqian</p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap site-overview-item animated">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
        <a href="/archives/">
          <span class="site-state-item-count">26</span>
          <span class="site-state-item-name">المقالات</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
        <span class="site-state-item-count">5</span>
        <span class="site-state-item-name">التصنيفات</span>
      </div>
      <div class="site-state-item site-state-tags">
        <span class="site-state-item-count">10</span>
        <span class="site-state-item-name">الوسوم</span>
      </div>
  </nav>
</div>



        </div>
      </div>
    </div>
  </aside>
  <div class="sidebar-dimmer"></div>


    </header>

    
  <div class="back-to-top" role="button" aria-label="Back to top">
    <i class="fa fa-arrow-up"></i>
    <span>0%</span>
  </div>

<noscript>
  <div class="noscript-warning">Theme NexT works best with JavaScript enabled</div>
</noscript>


    <div class="main-inner post posts-expand">


  


<div class="post-block">
  
  

  <article itemscope itemtype="http://schema.org/Article" class="post-content" lang="zh">
    <link itemprop="mainEntityOfPage" href="https://vampireqian.github.io/2017/09/22/11python%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86%E3%80%81%E6%A8%A1%E5%9D%97%E5%8C%96/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.gif">
      <meta itemprop="name" content="luoqian">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="luoqian">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          python-异常处理、模块化
        </h1>

        <div class="post-meta-container">
          <div class="post-meta">
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar"></i>
      </span>
      <span class="post-meta-item-text">نُشر في</span>

      <time title="أُنشأ: 2017-09-22 00:00:00" itemprop="dateCreated datePublished" datetime="2017-09-22T00:00:00+08:00">2017-09-22</time>
    </span>
      <span class="post-meta-item">
        <span class="post-meta-item-icon">
          <i class="far fa-calendar-check"></i>
        </span>
        <span class="post-meta-item-text">عُدل في</span>
        <time title="عُدل: 2018-12-11 16:36:38" itemprop="dateModified" datetime="2018-12-11T16:36:38+08:00">2018-12-11</time>
      </span>
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-folder"></i>
      </span>
      <span class="post-meta-item-text">في</span>
        <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
          <a href="/categories/python/" itemprop="url" rel="index"><span itemprop="name">python</span></a>
        </span>
    </span>

  
</div>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">
        <h2 id="异常处理"><a href="#异常处理" class="headerlink" title="异常处理"></a>异常处理</h2><h3 id="异常Exception"><a href="#异常Exception" class="headerlink" title="异常Exception"></a>异常Exception</h3><hr>
<p><strong>错误 Error</strong> ：错误是可以避免的<br>逻辑错误：算法写错了，加法写成了减法<br>笔误：变量名写错了，语法错误<br>函数或类使用错误，其实这也属于逻辑错误 </p>
<p><strong>异常 Exception</strong> ：异常不可能避免<br>本意就是意外情况<br>这有个前提，没有出现上面说的错误，也就是说程序写的没有问题，但是在某些情况下，会出现一些意外，导致程序无法正常的执行下去。<br>例如open函数操作一个文件，文件不存在，或者创建一个文件时已经存在了，或者访问一个网络文件，突然断网了，这就是异常，是个意外的情况。 </p>
<p><strong>错误和异常</strong><br>在高级编程语言中，一般都有错误和异常的概念，异常是可以捕获，并被处理的，但是错误是不能被捕获的。</p>
<h3 id="产生异常"><a href="#产生异常" class="headerlink" title="产生异常"></a>产生异常</h3><hr>
<p>产生：</p>
<ul>
<li>raise 语句显式的抛出异常</li>
<li>Python解释器自己检测到异常并引发它</li>
</ul>
<p>程序会在异常抛出的地方中断执行，如果不捕获，就会提前结束程序（其实是终止当前线程的执行）</p>
<p><strong>raise语句</strong><br>raise后什么都没有，表示抛出<strong>最近一个被激活</strong>的异常，如果没有，则抛类型异常。这种方式很少用 。</p>
<p>raise后要求应该是BaseException类的子类或实例，如果是类，将被无参实例化。</p>
<h3 id="异常类及继承层次"><a href="#异常类及继承层次" class="headerlink" title="异常类及继承层次"></a>异常类及继承层次</h3><hr>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Python异常的继承</span></span><br><span class="line"> </span><br><span class="line">BaseException</span><br><span class="line"> +-- SystemExit</span><br><span class="line"> +-- KeyboardInterrupt</span><br><span class="line"> +-- GeneratorExit</span><br><span class="line"> +-- Exception</span><br><span class="line">      +-- RuntimeError</span><br><span class="line">      |    +-- RecursionError</span><br><span class="line">      +-- MemoryError</span><br><span class="line">      +-- NameError</span><br><span class="line">      +-- StopIteration</span><br><span class="line">      +-- StopAsyncIteration</span><br><span class="line">      +-- ArithmeticError</span><br><span class="line">      |    +-- FloatingPointError</span><br><span class="line">      |    +-- OverflowError</span><br><span class="line">      |    +-- ZeroDivisionError</span><br><span class="line">      +-- LookupError</span><br><span class="line">      |    +-- IndexError</span><br><span class="line">      |    +-- KeyError</span><br><span class="line">      +-- SyntaxError</span><br><span class="line">      +-- OSError</span><br><span class="line">      |    +-- BlockingIOError</span><br><span class="line">      |    +-- ChildProcessError</span><br><span class="line">      |    +-- ConnectionError</span><br><span class="line">      |    |    +-- BrokenPipeError</span><br><span class="line">      |    |    +-- ConnectionAbortedError</span><br><span class="line">      |    |    +-- ConnectionRefusedError</span><br><span class="line">      |    |    +-- ConnectionResetError</span><br><span class="line">      |    +-- FileExistsError</span><br><span class="line">      |    +-- FileNotFoundError</span><br><span class="line">      |    +-- InterruptedError</span><br><span class="line">      |    +-- IsADirectoryError</span><br><span class="line">      |    +-- NotADirectoryError</span><br><span class="line">      |    +-- PermissionError</span><br><span class="line">      |    +-- ProcessLookupError</span><br><span class="line">      |    +-- TimeoutError</span><br></pre></td></tr></table></figure>

<h3 id="BaseException及子类"><a href="#BaseException及子类" class="headerlink" title="BaseException及子类"></a>BaseException及子类</h3><hr>
<p><strong>BaseException</strong><br>所有内建异常类的基类是BaseException</p>
<p><strong>SystemExit</strong><br>sys.exit(n)函数引发的异常，异常不捕获处理，就直接交给Python解释器，解释器退出。n=0,正常退出，n=1异常退出。</p>
<p>如果except语句捕获了该异常，则继续向后面执行，如果没有捕获住该异常SystemExit，解释器直接退出程序。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"> </span><br><span class="line"><span class="built_in">print</span>(<span class="string">&#x27;before&#x27;</span>)</span><br><span class="line">sys.exit(<span class="number">1</span>)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&#x27;SysExit&#x27;</span>)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&#x27;outer&#x27;</span>) <span class="comment"># 不执行</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 捕获这个异常</span></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">try</span>:</span><br><span class="line">    sys.exit(<span class="number">1</span>)</span><br><span class="line"><span class="keyword">except</span> SystemExit: <span class="comment"># 换成Exception</span></span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&#x27;SysExit&#x27;</span>)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&#x27;outer&#x27;</span>) <span class="comment"># 执行</span></span><br></pre></td></tr></table></figure>

<p><strong>KeyboardInterrupt</strong><br>对应的捕获用户中断行为Ctrl + C</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span>:</span><br><span class="line">    <span class="keyword">import</span> time</span><br><span class="line">    <span class="keyword">while</span> <span class="literal">True</span>:</span><br><span class="line">        time.sleep(<span class="number">0.5</span>)</span><br><span class="line">        <span class="keyword">pass</span></span><br><span class="line"><span class="keyword">except</span> KeyboardInterrupt:</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&#x27;ctl + c&#x27;</span>)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&#x27;outer&#x27;</span>)</span><br></pre></td></tr></table></figure>

<h3 id="Exception及子类"><a href="#Exception及子类" class="headerlink" title="Exception及子类"></a>Exception及子类</h3><p><strong>Exception</strong>是所有内建的、非系统退出的异常的基类，自定义异常应该继承自它</p>
<p><strong>SyntaxError</strong> 语法错误<br>Python将这种错误也归到异常类下面的Exception下的子类，但是这种错误是不可捕获的</p>
<p><strong>ArithmeticError</strong> 所有算术计算引发的异常，其子类有除零异常等</p>
<p><strong>LookupError</strong><br>使用映射的键或序列的索引无效时引发的异常的基类：IndexError, KeyError</p>
<p><strong>自定义异常</strong><br>从Exception继承的类</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="class"><span class="keyword">class</span> <span class="title">MyException</span>(<span class="params">Exception</span>):</span></span><br><span class="line">    <span class="keyword">pass</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">try</span>:</span><br><span class="line">    <span class="keyword">raise</span> MyException()</span><br><span class="line"><span class="keyword">except</span> MyException: <span class="comment"># 捕获自定义异常</span></span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&#x27;catch the exception&#x27;</span>)</span><br></pre></td></tr></table></figure>
<h3 id="未实现和未实现异常"><a href="#未实现和未实现异常" class="headerlink" title="未实现和未实现异常"></a>未实现和未实现异常</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">print</span>(<span class="built_in">type</span>(<span class="literal">NotImplemented</span>))</span><br><span class="line"><span class="built_in">print</span>(<span class="built_in">type</span>(NotImplementedError))</span><br><span class="line"></span><br><span class="line"><span class="comment">#&lt;class &#x27;NotImplementedType&#x27;&gt;</span></span><br><span class="line"><span class="comment">#&lt;class &#x27;type&#x27;&gt;</span></span><br></pre></td></tr></table></figure>

<p>NotImplemented是个值，单值，是NotImplementedType的实例</p>
<p>NotImplementedError是类型，是异常，返回type</p>
<h3 id="异常的捕获"><a href="#异常的捕获" class="headerlink" title="异常的捕获"></a>异常的捕获</h3><hr>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span>:</span><br><span class="line">    待捕获异常的代码块</span><br><span class="line"><span class="keyword">except</span> [异常类型]:</span><br><span class="line">    异常的处理代码块</span><br></pre></td></tr></table></figure>

<p>使用了<strong>try…except语句</strong>块捕捉到了这个异常，异常生成位置之后语句将不再执行，转而执行对应的except部分的语句，最后执行try…except语句块之外的语句。</p>
<p>except 后接异常类型，用来捕获指定类型的异常，except可以捕获多个异常。</p>
<p>捕获规则<br>捕获是从上到下依次比较，如果匹配，则执行匹配的except语句块<br>如果被一个except语句捕获，其他except语句就不会再次捕获了<br>如果没有任何一个except语句捕获到这个异常，则该异常向外抛出</p>
<p>捕获的原则<br>从小到大，从具体到宽泛</p>
<p>被抛出的异常，应该是异常的实例，使用as子句接收这个抛出的异常。</p>
<h3 id="finally子句"><a href="#finally子句" class="headerlink" title="finally子句"></a>finally子句</h3><hr>
<p>finally<br>最终，即最后一定要执行的，try…finally语句块中，不管是否发生了异常，都要执行finally的部分</p>
<p>finally中一般放置资源的清理、释放工作的语句，也可以在finally中再次捕捉异常。</p>
<h3 id="异常的传递"><a href="#异常的传递" class="headerlink" title="异常的传递"></a>异常的传递</h3><hr>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">foo1</span>():</span></span><br><span class="line">​    <span class="keyword">return</span> <span class="number">1</span>/<span class="number">0</span></span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">foo2</span>():</span></span><br><span class="line">​    <span class="built_in">print</span>(<span class="string">&#x27;foo2 start&#x27;</span>)</span><br><span class="line">​    foo1()</span><br><span class="line">​    <span class="built_in">print</span>(<span class="string">&#x27;foo2 stop&#x27;</span>)</span><br><span class="line"></span><br><span class="line">foo2()</span><br></pre></td></tr></table></figure>
<p>foo2调用了foo1，foo1产生的异常，传递到了foo2中。<br>异常总是向外层抛出，如果外层没有处理这个异常，就会继续向外抛出<br>如果内层捕获并处理了异常，外部就不能捕获到了<br>如果到了最外层还是没有被处理，就会中断异常所在的线程的执行。注意整个程序结束的状态返回值。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 线程中测试异常</span></span><br><span class="line"><span class="keyword">import</span> threading</span><br><span class="line"><span class="keyword">import</span> time</span><br><span class="line"> </span><br><span class="line"> </span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">foo1</span>():</span></span><br><span class="line">    <span class="keyword">return</span> <span class="number">1</span>/<span class="number">0</span></span><br><span class="line"> </span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">foo2</span>():</span></span><br><span class="line">    time.sleep(<span class="number">3</span>) <span class="comment"># 3秒后抛出异常</span></span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&#x27;foo2 start&#x27;</span>)</span><br><span class="line">    foo1()</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&#x27;foo2 stop&#x27;</span>)</span><br><span class="line"> </span><br><span class="line">t = threading.Thread(target=foo2)</span><br><span class="line">t.start()</span><br><span class="line"> </span><br><span class="line"><span class="keyword">while</span> <span class="literal">True</span>:</span><br><span class="line">    time.sleep(<span class="number">1</span>)</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&#x27;Everything OK&#x27;</span>)</span><br><span class="line">    <span class="keyword">if</span> t.is_alive():</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">&#x27;alive&#x27;</span>)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">&#x27;dead&#x27;</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>




<h3 id="try嵌套"><a href="#try嵌套" class="headerlink" title="try嵌套"></a>try嵌套</h3><hr>
<p>内部捕获不到异常，会向外层传递异常<br>但是如果内层有finally且其中有return、break语句，则异常就不会继续向外抛出：异常被压制。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span>:</span><br><span class="line">    <span class="keyword">try</span>:</span><br><span class="line">        ret = <span class="number">1</span> / <span class="number">0</span></span><br><span class="line">    <span class="keyword">except</span> KeyError <span class="keyword">as</span> e:</span><br><span class="line">        <span class="built_in">print</span>(e)</span><br><span class="line">    <span class="keyword">finally</span>:</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">&#x27;inner fin&#x27;</span>)</span><br><span class="line"><span class="keyword">except</span>:</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&#x27;outer catch&#x27;</span>)</span><br><span class="line"><span class="keyword">finally</span>:</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&#x27;outer fin&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#输出    </span></span><br><span class="line"><span class="comment">#inner fin</span></span><br><span class="line"><span class="comment">#outer catch</span></span><br><span class="line"><span class="comment">#outer fin</span></span><br></pre></td></tr></table></figure>

<h3 id="异常的捕获的时机"><a href="#异常的捕获的时机" class="headerlink" title="异常的捕获的时机"></a>异常的捕获的时机</h3><p>1.立即捕获</p>
<p>需要立即返回一个明确的结果</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">parse_int</span>(<span class="params">s</span>):</span></span><br><span class="line">    <span class="keyword">try</span>:</span><br><span class="line">        <span class="keyword">return</span> <span class="built_in">int</span>(s)</span><br><span class="line">    <span class="keyword">except</span>:</span><br><span class="line">        <span class="keyword">return</span> <span class="number">0</span></span><br><span class="line"> </span><br><span class="line"><span class="built_in">print</span>(parse_int(<span class="string">&#x27;s&#x27;</span>))</span><br></pre></td></tr></table></figure>

<p>2.边界捕获</p>
<p>封装产生了边界</p>
<p>例如，写了一个模块，用户调用这个模块的时候捕获异常，模块内部不需要捕获、处理异常，一旦内部处理了，外<br>部调用者就无法感知了。<br>例如，open函数，出现的异常交给调用者处理，文件存在了，就不用再创建了，看是否修改还是删除<br>例如，自己写了一个类，使用了open函数，但是出现了异常不知道如何处理，就继续向外层抛出，一般来说最外<br>层也是边界，必须处理这个异常了，否则线程退出</p>
<h3 id="else子句"><a href="#else子句" class="headerlink" title="else子句"></a>else子句</h3><figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span>:</span><br><span class="line">    ret = <span class="number">1</span> * <span class="number">0</span></span><br><span class="line"><span class="keyword">except</span> ArithmeticError <span class="keyword">as</span> e:</span><br><span class="line">    <span class="built_in">print</span>(e)</span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&#x27;OK&#x27;</span>)</span><br><span class="line"><span class="keyword">finally</span>:</span><br><span class="line">    <span class="built_in">print</span>(<span class="string">&#x27;fin&#x27;</span>)</span><br></pre></td></tr></table></figure>

<p>else子句<br><strong>没有任何异常发生</strong>，则执行</p>
<h3 id="总结"><a href="#总结" class="headerlink" title="总结"></a>总结</h3><hr>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">try</span>:</span><br><span class="line">    &lt;语句&gt;    <span class="comment">#运行别的代码</span></span><br><span class="line"><span class="keyword">except</span> &lt;异常类&gt;：</span><br><span class="line">    &lt;语句&gt;    <span class="comment"># 捕获某种类型的异常</span></span><br><span class="line"><span class="keyword">except</span> &lt;异常类&gt; <span class="keyword">as</span> &lt;变量名&gt;:</span><br><span class="line">    &lt;语句&gt;    <span class="comment"># 捕获某种类型的异常并获得对象</span></span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    &lt;语句&gt;    <span class="comment">#如果没有异常发生</span></span><br><span class="line"><span class="keyword">finally</span>:</span><br><span class="line">    &lt;语句&gt;    <span class="comment">#退出try时总会执行</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><strong>try的工作原理</strong> </p>
<p>1、如果try中语句执行时发生异常，搜索except子句，并执行第一个匹配该异常的except子句<br>2、如果try中语句执行时发生异常，却没有匹配的except子句，异常将被递交到外层的try，如果外层不处理这个异常，异常将继续向外层传递。如果都不处理该异常，则会传递到最外层，如果还没有处理，就终止异常所在的线程<br>3、如果在try执行时没有发生异常，将执行else子句中的语句<br>4、无论try中是否发生异常，finally子句最终都会执行。</p>
<h2 id="模块化"><a href="#模块化" class="headerlink" title="模块化"></a>模块化</h2><p>Python中只有一种模块对象类型，但是为了模块化组织模块的便利，提供了“包”的概念。模块module，指的是Python的源代码文件。</p>
<p>包package，指的是模块组织在一起的和包名同名的目录及其相关文件。</p>
<h3 id="导入语句"><a href="#导入语句" class="headerlink" title="导入语句"></a>导入语句</h3><table>
<thead>
<tr>
<th>语句</th>
<th>含义</th>
</tr>
</thead>
<tbody><tr>
<td>import 模块1，[模块2….]</td>
<td>完全导入</td>
</tr>
<tr>
<td>import….as…..</td>
<td>模块别名</td>
</tr>
</tbody></table>
<p><strong>import语句</strong></p>
<p>1、找到指定的模块，加载和初始化它，生成模块对象。找不到，抛出ImportError</p>
<p>2、在import所在的作用域的局部命名空间中，增加名称和上一步创建的对象关联。</p>
<p><strong>总结</strong></p>
<p>导入顶级模块，其名称会加入到本地名词空间中，并绑定到其模块对象。</p>
<p>导入非顶层模块，只将其顶级模块名称加入到本地名称空间中。导入的模块必须使用完全限定名称来访问。</p>
<p>如果使用了as，as后的名称直接绑定到导入的模块对象，并将该名称加入到本地名词空间中。</p>

    </div>

    
    
    

    <footer class="post-footer">
          <div class="post-tags">
              <a href="/tags/python/" rel="tag"># python</a>
          </div>

        

          <div class="post-nav">
            <div class="post-nav-item">
                <a href="/2017/09/10/09python%E6%96%87%E4%BB%B6%E5%A4%84%E7%90%86%E3%80%81%E8%B7%AF%E5%BE%84%E5%A4%84%E7%90%86%E3%80%81%E5%BA%8F%E5%88%97%E5%8C%96%E5%92%8C%E5%8F%8D%E5%BA%8F%E5%88%97%E5%8C%96/" rel="prev" title="python-文件处理、路径处理、序列化和反序列化">
                  <i class="fa fa-chevron-left"></i> python-文件处理、路径处理、序列化和反序列化
                </a>
            </div>
            <div class="post-nav-item">
                <a href="/2017/09/30/12python%E7%9B%B4%E6%8E%A5%E8%B5%8B%E5%80%BC%E3%80%81%E6%B7%B1%E6%8B%B7%E8%B4%9D%E5%92%8C%E6%B5%85%E6%8B%B7%E8%B4%9D/" rel="next" title="python直接赋值、深拷贝和浅拷贝">
                  python直接赋值、深拷贝和浅拷贝 <i class="fa fa-chevron-right"></i>
                </a>
            </div>
          </div>
    </footer>
  </article>
</div>






</div>
  </main>

  <footer class="footer">
    <div class="footer-inner">


<div class="copyright">
  &copy; 
  <span itemprop="copyrightYear">2021</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">luoqian</span>
</div>
  <div class="powered-by">تطبيق الموقع <a href="https://hexo.io/" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.js.org/muse/" rel="noopener" target="_blank">NexT.Muse</a>
  </div>

    </div>
  </footer>

  
  <script src="https://cdn.jsdelivr.net/npm/animejs@3.2.1/lib/anime.min.js" integrity="sha256-XL2inqUJaslATFnHdJOi9GfQ60on8Wx1C2H8DYiN1xY=" crossorigin="anonymous"></script>
<script src="/js/comments.js"></script><script src="/js/utils.js"></script><script src="/js/motion.js"></script><script src="/js/schemes/muse.js"></script><script src="/js/next-boot.js"></script>

  





  





</body>
</html>
